src/qtpromise/qpromise.h

```diff
@@ -71,9 +75,22 @@ class QPromiseBase
     inline QPromise<T> tapFail(THandler handler) const;
 
     template <typename E = QPromiseTimeoutException>
-    inline QPromise<T> timeout(int msec, E&& error = E()) const;
+    inline QPromise<T> timeout(int msec, E&& error = E()) const;   
+#if __cplusplus >= 201103
+    template <typename E = QPromiseTimeoutException>
+    inline QPromise<T> timeout(std::chrono::milliseconds msec, E&& error = E()) const
+    {
+        return timeout(static_cast<int>(msec.count()), std::forward<E>(error));
```

Comment:
I was also tending towards it but was not sure if you require std::chrono to be implemented by the compiler. At least the Qt Framework itself requires a C++-compliant compiler but says that the standard library is optional

Reply:
`QTimer` uses `__has_include`, which kind makes more sense than checking for `__cplusplus >= 201103`. But is it to support compilers prior c++11 or is it needed because some c++11 compilers don't support `std::chrono`? I can't find docs about it.

Reply:
Well, this is where it gets interesting. 

I've passed a similar patch to Qt 5.15 recently, and I was told to use `__has_include` to detect whether `<chrono>` is present or not. According to [1], `__has_include` is a C++17 feature  but my patch failed to compile on MacOS in Qt's pipeline because I used `std::chrono_literals`, which is a C++14 standard library. So I removed `std::chrono_literals` but left `__has_include` intact, and it went through.

My first implementation of this PR to QtPromise had used `__has_include` but failed to compile with msvc2013. I was looking for a good way to detect whether `<chrono>` is available and I stumbled upon `__cpp_lib_chrono` as in [2] but it seems to be a C++14 feature. So I ended up checking the `__cplusplus` but as you already mentioned it might be useless. 

There was a problem with the INTEGRITY compiler in the CI pipeline of Qt not so long ago, and it was said in the mailing list that Qt requires a C++11 compiler but not the standard library [3]. At the Qt World Summit last year I was told that **some** of standard library is required but not all of it. So there **might** be a user of the Qt framework using your library and not having the C++ standard library.

So I think we can do one of these: 
1. Require `<chrono>` from all users of QtPromise. I personally think this is a safe choice and would prefer it because I don't care about more exotic toolchains like INTEGRITY but you may be of other opinion.
2. Test compiler versions using compiler-specific macros like _MSC_VER, and then fall back to `__has_include`. If there is no `__has_include`, assume there is no `<chrono>`. gcc 5 has already supported `__has_include`, according to [4]. Unfortunately, I don't have an old enough msvc to test. 

I can also look into the sources of Qt 5.6.0 since this is the version you support, and check whether there is a better solution there. 

[1] https://en.cppreference.com/w/cpp/preprocessor/include
[2] https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros
[3] https://lists.qt-project.org/pipermail/development/2019-September/037479.html
[4] https://gcc.gnu.org/projects/cxx-status.html

Reply:
Thank you for the detailed answer. I realize that my previous comment doesn't make sense: it's not about compiler capabilities but the STL. So checking for c++11 compat doesn't really solve the problem because it doesn't guarantee that the `<chrono>` header is present. I would not rely on `__has_include` either since it's a c++17 feature: I'm curious to know how Qt can use it and still be compatible with projects not supporting c++17.

Since we are not testing on such toolchains, I'm not sure we should try to handle these cases until they are reported. Also, to be consistent, we would need to check presence of `<functional>` and `<array>`, right?

So I'm totally fine with your first proposal to make `<chrono>` required. And in this case, I would keep your current approach to make the `std::chrono::milliseconds msec` fallback to the `int msec` method (and just move the implementation in the `.inl` file).

Reply:
I've looked into the source of Qt 5.8 (the first version where std::chrono overloads appeared in QTimer), they use a `QT_HAS_INCLUDE` macro which turns into `__has_include` on compilers that support it, and to 0 otherwise. So I think that using std::chrono in QTimer is not even possible with MSVC2013 but, as said, I can't really test it. 

As for __has_include, although this feature is a C++17 requirement, it has been supported by gcc since at least 5 series (released in 2015). 

Regarding `<functional>` and `<array>`: If your library inherently depends on them, they should be required. The `chrono` overloads are a mere convenience, so `<chrono>` might as well be optional. But, again, I think the diversity of platforms that Qt runs on has much more implication on the compiler requirements. Let's say somebody is porting QtPromise on a microcontroller where there is no C++ standard library, but only a C++11-compliant compiler. They would definitely want a bare minimum of dependencies.